invenio_records_resources/services/records/results.py

```diff
@@ -352,12 +362,26 @@ def _find_fields(self, service, value):
 
     def _fetch_referenced(self, grouped_values, identity):
         """Search and fetch referenced recs by ids."""
+
+        def _add_dereferenced_record(service, value, resolved_rec):
+            """Helper function to set the dereferenced record to the service."""
+            for field in self._find_fields(service, value):
+                field.add_dereferenced_record(service, value, resolved_rec)
+
         for service, values in grouped_values.items():
```

Comment:
There is a lot of nested looping in this function and it is hard to understand, maybe adding more comments would help.